microsoft/testsuites/cloud_hypervisor/ch_tests_tool.py

```diff
@@ -49,35 +51,84 @@ def run_tests(
+                force_run=True,
+                cwd=self.repo_root,
+                no_info_log=False,  # print out result of each test
+                shell=True,
             )
 
-        assert_that(failures, f"Unexpected failures: {failures}").is_empty()
+            results = self._extract_test_results(result.stdout)
+            failures = [r.name for r in results if r.status == TestStatus.FAILED]
+            if not failures:
+                result.assert_exit_code()
+
+            for r in results:
+                self._send_subtest_msg(
+                    test_result.id_,
+                    environment,
+                    r.name,
+                    r.status,
+                )
+
+            assert_that(failures, f"Unexpected failures: {failures}").is_empty()
+        else:
+            temp = self.repo_root.joinpath("perf_mtr_report")
+            self.per_mtr_report_file = temp.joinpath("testcase_result.json")
+
+            self._build_ch_with_mshv(hypervisor=hypervisor)
+            perf_metrics_tests = self._list_perf_metrics_tests(hypervisor=hypervisor)
+            testcase_result = {}
+            for testcase in perf_metrics_tests:
+                self._log.info(f"Running testcase : {testcase}")
+                testcase_result[testcase] = {}
+                testcase_result[testcase]["status"] = "Started"
+
+                try:
+                    result = self.run(
+                        f"tests --hypervisor {hypervisor} --metrics -- -- \
+                            --test-filter {testcase}",
+                        timeout=self.TIME_OUT,
+                        force_run=True,
+                        cwd=self.repo_root,
+                        no_info_log=False,  # print out result of each test
+                        shell=True,
+                    )
+                    output = result.stdout.replace("\r\n", "\n")
+                    output = output.replace("\t", "")
+                    if result.exit_code != 0:
+                        excep = Exception(f"Test Case Failed : {testcase}", output)
+                        raise excep
+                    testcase_result[testcase]["status"] = "Succeeded"
+                    metrics = self._process_perf_metric_test_result(result.stdout)
+                    testcase_result[testcase]["metrics"] = metrics
+                    testcase_result[testcase]["err"] = ""
+                except Exception as e:
+                    self._log.error(f"Testcase failed , tescase name : {testcase}")
+                    testcase_result[testcase]["status"] = "Failed"
+                    testcase_result[testcase]["err"] = str(e.args[0])
+                    testcase_result[testcase]["trace"] = str(e.args[1])
+                    testcase_result[testcase]["metrics"] = ""
+            self._log.debug(testcase_result)
+            report = self._create_perf_metric_report(testcase_result)
+
+            for record in report:
+                self._send_subtest_msg(
```

Comment:
Include error information into the subtest messages.

Reply:
sure Chi, i will look for this.

Reply:
Added test message in subtest. It will be error if testcase failed or metrics in case of success case.